src/nbl/video/device_limits.json

```diff
@@ -0,0 +1,2751 @@
+            {
+                "type": "uint32_t",
+                "name": "maxOptimallyResidentWorkgroupInvocations",
+                "value": 0,
+                "comment": "its 1D because multidimensional workgroups are an illusion"
+            },
+            {
+                "type": "uint32_t",
+                "name": "maxResidentInvocations",
+                "value": 0,
+                "comment": "These are maximum number of invocations you could expect to execute simultaneously on this device"
+            },
+            {
+                "type": "asset::CGLSLCompiler::E_SPIRV_VERSION",
+                "name": "spirvVersion",
+                "value": "asset::CGLSLCompiler::E_SPIRV_VERSION::ESV_1_8"
+            },
+            {
+                "type": "function",
+                "function": [
+                    "// utility functions",
+                    "// In the cases where the workgroups synchronise with each other such as work DAGs (i.e. `CScanner`),",
+                    "// `workgroupSpinningProtection` is meant to protect against launching a dispatch so wide that",
+                    "// a workgroup of the next cut of the DAG spins for an extended time to wait on a workgroup from a previous one.",
+                    "inline uint32_t computeOptimalPersistentWorkgroupDispatchSize(const uint64_t elementCount, const uint32_t workgroupSize, const uint32_t workgroupSpinningProtection=1u) const",
+                    "{",
+                    "    assert(elementCount!=0ull && \"Input element count can't be 0!\");",
+                    "    const uint64_t infinitelyWideDeviceWGCount = (elementCount-1ull)/(static_cast<uint64_t>(workgroupSize)*static_cast<uint64_t>(workgroupSpinningProtection))+1ull;",
+                    "    const uint32_t maxResidentWorkgroups = maxResidentInvocations/workgroupSize;",
+                    "    return static_cast<uint32_t>(core::min<uint64_t>(infinitelyWideDeviceWGCount,maxResidentWorkgroups));",
+                    "}"
+                ]
+            },
+            {
+                "type": "function",
+                "function": [
+                    "// we don't compare certain capabilities because they don't mean better/worse",
+                    "inline bool isSubsetOf(const SPhysicalDeviceLimits& _rhs) const",
+                    "{",
+                    "    if (maxImageDimension1D > _rhs.maxImageDimension1D) return false;",
+                    "    if (maxImageDimension2D > _rhs.maxImageDimension2D) return false;",
+                    "    if (maxImageDimension3D > _rhs.maxImageDimension3D) return false;",
+                    "    if (maxImageDimensionCube > _rhs.maxImageDimensionCube) return false;",
+                    "    if (maxImageArrayLayers > _rhs.maxImageArrayLayers) return false;",
+                    "    if (maxBufferViewTexels > _rhs.maxBufferViewTexels) return false;",
+                    "    if (maxUBOSize > _rhs.maxUBOSize) return false;",
```

Comment:
this should be missing from the JSON, and the body should be autogenerated by python by just knowing the type of the member variable you declared above.

If you manually neeed to maintain this then we're back to square 1 this becomes a handwritten header but worse